backends/candle/src/models/flash_modernbert.rs

```diff
@@ -0,0 +1,483 @@
+
+        let attn_outputs = self.attn.forward(&attn_norm, cu_seqlens, cos, sin, max_s)?;
+
+        let hidden_states = residual.add(&attn_outputs)?;
+
+        let mlp_output = self
+            .mlp
+            .forward(&self.mlp_norm.forward(&hidden_states, None)?)?;
+
+        hidden_states.add(&mlp_output)
+    }
+}
+
+struct ModernBertEncoder {
+    layers: Vec<ModernBertEncoderLayer>,
+
+    global_attn_every_n_layers: usize,
+
+    span: tracing::Span,
+}
+
+impl ModernBertEncoder {
+    pub fn load(vb: VarBuilder, config: &ModernBertConfig) -> Result<Self> {
+        let layers = (0..config.num_hidden_layers)
+            .map(|index| ModernBertEncoderLayer::load(vb.pp(format!("{index}")), index, config))
+            .collect::<Result<Vec<_>>>()?;
+
+        let span = tracing::span!(tracing::Level::TRACE, "encoder");
+
+        Ok(ModernBertEncoder {
+            layers,
+            global_attn_every_n_layers: config.global_attn_every_n_layers,
+            span,
+        })
+    }
+
+    fn forward(
+        &self,
+        hidden_states: &Tensor,
+        cu_seqlens: &Tensor,
+        rotary_cache: &HashMap<bool, (Tensor, Tensor)>,
+        max_s: usize,
+    ) -> Result<Tensor> {
+        let _enter = self.span.enter();
+
+        let mut hidden_states = hidden_states.clone();
+
+        for (index, layer) in self.layers.iter().enumerate() {
+            let use_local_attention = index % self.global_attn_every_n_layers != 0;
+            let (cos, sin) = &rotary_cache[&use_local_attention];
+
+            hidden_states = layer.forward(&hidden_states, cu_seqlens, cos, sin, max_s)?;
+        }
+
+        Ok(hidden_states)
+    }
+}
+
+pub struct FlashModernBertModel {
+    embeddings: ModernBertEmbeddings,
+    encoder: ModernBertEncoder,
+    final_norm: LayerNorm,
+    pool: Pool,
+    classifier: Option<Box<dyn ClassificationHead + Send>>,
+
+    rotary_dim: usize,
+    rotary_cache: HashMap<bool, (Tensor, Tensor)>,
+
+    device: Device,
+
+    span: tracing::Span,
+}
+
+impl FlashModernBertModel {
+    pub fn load(vb: VarBuilder, config: &ModernBertConfig, model_type: ModelType) -> Result<Self> {
+        match vb.device() {
+            Device::Cuda(_) => {}
+            _ => candle::bail!("FlashModernBert requires Cuda"),
+        }
+
+        if vb.dtype() != DType::F16 {
+            candle::bail!("FlashModernBert requires DType::F16")
+        }
+
+        let (pool, classifier) = match model_type {
+            ModelType::Classifier => {
+                let pool = Pool::Cls;
+
+                let classifier: Box<dyn ClassificationHead + Send> =
+                    Box::new(ModernBertClassificationHead::load(vb.clone(), config)?);
+
+                (pool, Some(classifier))
+            }
+            ModelType::Embedding(pool) => {
+                if pool == Pool::Splade {
+                    candle::bail!("`splade` is not supported for ModernBert")
+                }
+
+                if pool == Pool::LastToken {
```

Comment:
Should be implemented below, right?

Reply:
thanks for pointing this! I had mistakenly disabled support for LastToken pooling, even though it was already implemented. I've removed the line blocking its support, allowing LastToken pooling to be used again.

https://github.com/huggingface/text-embeddings-inference/pull/459/commits/1fe761f779ac6f465011b05513786e6e11ed4ce9